designs/2018-relative-package-loading/README.md

```diff
@@ -0,0 +1,278 @@
+
+Duplicate reports seem to be a necessary side-effect of the goal to allow different plugins with the same name to be configured simultaneously. For example, if `eslint-config-foo` and `eslint-config-bar` had different configs for `react/no-typos`, then running both rules simultaneously might be a desirable behavior. (Consider a rule like [`no-restricted-syntax`](https://eslint.org/docs/rules/no-restricted-syntax), which can report entirely different code patterns depending on configuration.) Even if `eslint-config-foo` and `eslint-config-bar` used the same configuration for `react/no-typos`, running both versions of the rule simultaneously might still be desirable if the different versions of `eslint-plugin-react` had substantially different behaviors for the `no-typos` rule.
+
+In the case where `eslint-config-foo` and `eslint-config-bar` both end up loading the exact same copy of `eslint-plugin-react` (e.g. due to a package manager's tree flattening), ESLint could avoid the performance cost of running the same rule twice by simply running the rule once and producing two reports. (It doesn't seem like a good idea to only output one report in this case, because this would make ESLint's output highly dependent on how a package manager decides to arrange packages in `node_modules`.)
+
+The UI impact of this problem could largely be mitigated by updating formatters to deduplicate reports when displaying them. For example, rather than displaying two separate lines for two reports with the same message at the same location, a formatter could display the message once and list both rule names.
+
+### Shareable configs can no longer reconfigure their siblings' plugins
+
+If a user extends `eslint-config-foo` and `eslint-config-bar`, and `eslint-config-foo` depends on `eslint-plugin-react`, then `eslint-config-bar` cannot reference the version of `eslint-plugin-react` used by `eslint-config-foo`. (If `eslint-config-bar` wanted to use `eslint-plugin-react`, it could instead install its own version of `eslint-plugin-react`, independently of `eslint-config-foo`.)
+
+This property is by design; since `eslint-config-bar` can't know what version of `eslint-plugin-react` is used by its siblings (or if `eslint-plugin-react` is in use at all), it generally wouldn't know how to produce a reasonable configuration for the rules in `eslint-plugin-react`. However, this will break some aspects of shareable configs like [`eslint-config-prettier`](https://github.com/prettier/eslint-config-prettier), which attempts to disable all stylistic rules from a set of a few popular plugins. (The particular issue of disabling stylistic rules might be solvable separately from this proposal since rules can now be explicitly marked as stylistic through the `type` metadata flag.)
```

Comment:
To clarify, this would only break the plugin-specific configs bundled in `eslint-config-prettier` (namely `prettier/flowtype`, `prettier/react`, `prettier/standard`, `prettier/unicorn`, and `prettier/vue`), not the main config which addresses core rules. I don't have empirical data about the usage of these configs, but I suspect their usage is significantly smaller. (In any case, I agree we shouldn't take the breakage lightly.)

As part of the discussion on adding the `--fix-type` flag, we [also considered the possibility that it could be used to replace `eslint-config-prettier`](https://gitter.im/eslint/tsc-meetings?at=5bad40ed691dc567dbb340c5). I think an ideal solution would be to provide a migration path by adding something like that to core.

Other possible migration paths include (ranked roughly in order of preference, although they aren't mutually exclusive):

* Plugins could expose configs that disable all of their stylistic rules. Then a user could use something like `extends: ['airbnb', 'plugin:airbnb::react/disable-all-stylistic-rules']` to disable all the stylistic rules that were originally enabled by `eslint-config-airbnb`.
* Using wrappers around shareable configs that disable stylistic rules. The use case for `eslint-config-prettier` is usually that stylistic rules have been enabled by another shareable config (e.g. `eslint-config-airbnb`), and the user wants to disable all of those stylistic rules. So instead of enumerating rules from popular plugins, `eslint-config-prettier` could instead expose configs that wrap popular shareable configs, e.g. a `prettier/airbnb` config which is the same as `eslint-config-airbnb` except that it disables all stylistic rules. Then an end user would use `the prettier/airbnb` config as a substitute for the `airbnb` config.
* Using scripting outside of the ESLint config extension process. For example, a package could expose a function that accepts the end user's config object and adds `foo: "off"` to the rules section (substituting the appropriate rule names for `"foo"`). This could lead to ambiguity errors if multiple version of a plugin were installed, so it wouldn't be an ideal long-term solution, but it would provide a migration path for existing cases.
* In ESLint core, we could allow something like `..` in a config scope to refer to a parent in the tree (which would allow `eslint-config-prettier` to use something like `..::react` to refer to its siblings). But this seems like it would break some invariants of the design and could result fragile cases, such as configs that depend on being used as siblings of other configs. This solution doesn't seem worthwhile just for compatibility with a single shareable config.